src/openms/source/QC/Contaminants.cpp

```diff
@@ -35,113 +35,181 @@
   {
-    String enzyme = features.getProteinIdentifications()[0].getSearchParameters().digestion_enzyme.getName();
-    if (enzyme == "unknown_enzyme")
+    //empty FeatureMap
+    if (features.empty())
     {
-      LOG_WARN << "No digestion enzyme in featureMap detected.";
+      LOG_WARN << "FeatureMap is empty." << "\n";
+      return;
     }
-    ProteaseDigestion digestor;
-    digestor.setEnzyme(enzyme);
-    UInt missed_cleavages (features.getProteinIdentifications()[0].getSearchParameters().missed_cleavages);
-    digestor.setMissedCleavages(missed_cleavages);
-
-    Size dropped_by_length(0); // stats for removing candidates
-
-
-    for (const FASTAFile::FASTAEntry& fe : contaminants)
+    //empty features
+    if (features[0].getPeptideIdentifications().empty())
+    {
+      LOG_WARN << "Features are empty." << "\n";
+      return;
+    }
+    //empty contaminants database
+    if (contaminants.empty())
+    {
+      LOG_WARN << "FASTAFile is empty." << "\n";
+      return;
+    }
+    //fill the unordered set once with the digested contaminants database
+    if (digested_db_.empty())
     {
+      ProteaseDigestion digestor;
+      String enzyme = features.getProteinIdentifications()[0].getSearchParameters().digestion_enzyme.getName();
 
-      vector<AASequence> current_digest;
-      if (enzyme == "unknown_enzyme")                                                   //no enzyme given
+      //no enzyme is given
+      if (enzyme == "unknown_enzyme")
       {
-        current_digest.push_back(AASequence::fromString(fe.sequence));
+        LOG_WARN << "No digestion enzyme in featureMap detected." << "\n";
+        //contaminants database will not be digested
       }
-      else                                                                    //digest the sequence
+      else
       {
-        dropped_by_length += digestor.digest(AASequence::fromString(fe.sequence), current_digest);
+        digestor.setEnzyme(enzyme);
       }
-//      DigestedProtein temp;
-//      temp.id = fe.identifier;
-      for (auto const& s : current_digest)                                    //fill digested_db_ with digested sequences
+      //Get the missed cleavages for the digestor. If none are given, its default is 0.
+      UInt missed_cleavages(features.getProteinIdentifications()[0].getSearchParameters().missed_cleavages);
+      digestor.setMissedCleavages(missed_cleavages);
+
+      // stats for removing candidates
+      Size dropped_by_length(0);
+
+
+      //digest the contaminants database and add the peptides into the unordered set
+      for (const FASTAFile::FASTAEntry &fe : contaminants)
       {
-         digested_db_.insert(s.toUnmodifiedString());
-          //temp.peptides.push_back(s.toString());
+        vector<AASequence> current_digest;
+        //No digestion enzyme is given. Contaminant database will not get digested.
+        if (enzyme == "unknown_enzyme")
```

Comment:
not required anymore?

Reply:
no, the temp object does not exist anymore 
to save the strings we are using the unordered set instead